### PR TITLE
configure.ac: make -W cflag checks more robust

### PR DESCRIPTION
**Pull request by thac0**
Stumbled upon this while attempting to compile the codebase
(unsuccessfully) with clang. As it happens, the underlying autoconf
macro (AX_CHECK_COMPILE_FLAG) will only fail if the compiler returns
non-zero. After a few moments of exploration, this is not always
sufficient.

In the case of clang, an unrecognized cflag will emit a warning, but
clang will return 0. Therefore, configure will (incorrectly) conclude
the cflag is recognized. See #2097 for the result.

In the case of gcc, if a cflag is recognized but used incorrectly
(example: -Wformat-security by itself), gcc will emit a warning but
return 0.

So, add smarts (EXTRA-FLAGS) to the macro, which results in clang
or gcc exiting with an error should it encounter a cflag it does not
recognize or a cflag used incorrectly.

Signed-off-by: Joe Konno <joe.konno@intel.com>

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -225,12 +226,12 @@ AS_IF([test x"$enable_hardening" != x"no"], [
...
   ])
 
+  add_hardened_c_flag([-Werror])
   add_hardened_c_flag([-Wall])
   add_hardened_c_flag([-Wextra])
-  add_hardened_c_flag([-Werror])
 
   add_hardened_c_flag([-Wformat])
-  add_hardened_c_flag([-Wformat-security])
+  add_hardened_c_flag([-Wformat -Wformat-security])
```
Comment (williamcroberts) [inline configure.ac:234]: Why this change?

Comment (thac0) [reply configure.ac:234]: > Why this change?

Because, if `gcc -Wformat-security` is run by itself, you get:
```
$ gcc -Wformat-security -o foo foo.c
cc1: warning: ‘-Wformat-security’ ignored without ‘-Wformat’ [-Wformat-security]
```
While it is technically 'recognized', it would error out with -Werror flipped:
```
$ gcc -Werror -Wformat-security -o foo foo.c
cc1: error: ‘-Wformat-security’ ignored without ‘-Wformat’ [-Werror=format-security]
cc1: all warnings being treated as errors
```
So, in my mind, checking for this particular flag's existence on its own is invalid.

**Comment by tstellar** [comment]
I filed a very similar pull request in #1993 , but closed it in favor of #2075.  It seems to me that trying to define a set of hardening flags that work for clang might work better, but I am happy with any solution that will fix the clang build with hardening.

File: configure.ac [hunk 2]
```diff
@@ -207,7 +207,8 @@ AS_IF([test x"$enable_hardening" != x"no"], [
   AC_DEFUN([add_hardened_c_flag], [
     AX_CHECK_COMPILE_FLAG([$1],
       [EXTRA_CFLAGS="$EXTRA_CFLAGS $1"],
-      [AC_MSG_ERROR([Cannot enable $1, consider configuring with --disable-hardening])]
+      [AC_MSG_ERROR([Cannot enable $1, consider configuring with --disable-hardening])],
```
Comment (williamcroberts) [inline configure.ac:210]: Something else is still broken:
```
configure: WARNING: Compiling with --disable-hardening is dangerous!
you should consider fixing the configure script compiler flags
and submitting patches upstream!
checking whether C compiler accepts -D_GNU_SOURCE... yes
checking whether C compiler accepts -Wstringop-overflow=4... no
checking whether C compiler accepts -Wstringop-truncation... no
checking whether C compiler accepts -Wduplicated-branches... no
checking whether C compiler accepts -Wduplicated-cond... no
checking whether C compiler accepts -Wbool-compare... no
configure: error: Cannot enable -Wbool-compare, consider configuring with --disable-hardening
The command "./.ci/travis.run" exited with 1.
```
Perhaps --disable-hardening configure flag is not working?

File: configure.ac [hunk 3]
```diff
@@ -207,7 +207,8 @@ AS_IF([test x"$enable_hardening" != x"no"], [
...
+      [-Werror]
```
Comment (thac0) [inline configure.ac:211]: I think the reason Travis et al are failing, @williamcroberts , is because of this line. I just ran `gcc -Wbool` for grins, and it's an invalid flag, so with this change to the macro, it's now an error at configure-time, and therefore a show-stopper.

File: configure.ac (hunk 2 above)
Comment (thac0) [reply configure.ac:210]: See my comment on line 211

**Comment by thac0** [comment]
Yeah, @tstellar , I think we're missing the forest for the trees in how cflags are discovered. The fact that you have to `--disable-hardening` feels wrong to me, given clang's feature set. I'm cooking up an idea, so expect a PR relatively soon.

**Comment by thac0** [comment]
Looks like Cirrus failed because it didn't pull my modified & rebased change? Huh.

**Comment by williamcroberts** [comment]
I would:
1. Make it best attempt to add a default set
2. If you don't want to attempt to add the default set, configure option --disable-defaultflags to match tpm2-tss
    
    - I would like to, when possible, merge behaviors between projects.

**Comment by thac0** [comment]
I'm with you on all points, @williamcroberts . I was thinking hardest about the "default set" that would allow for easy building with gcc or clang. And if folks like it here, no reason why I couldn't apply that set/functionality to the other tpm projects.

In going through the motions, this PR has become less a proposed fix and more a discovery mission. So feel free to close if you like.

**Comment by williamcroberts** [comment]
> I'm with you on all points, @williamcroberts . I was thinking hardest about the "default set" that would allow for easy building with gcc or clang. And if folks like it here, no reason why I couldn't apply that set/functionality to the other tpm projects.
> 
> In going through the motions, this PR has become less a proposed fix and more a discovery mission. So feel free to close if you like.

I'll let you close it or you can force push the branch with the newest changes when you're ready. Thanks for taking the time to look into this.

**Comment by thac0** [comment]
Closing to pursue a more robust approach. Stay tuned for a new PR.
